File: solutions/1306. Jump Game III.java
class Solution {
    private int[] arr;
    private Set<Integer> visited = new HashSet<>();
    
    public boolean canReach(int[] nums, int start) {
        arr = nums;
        boolean ans = isPossibleToReach(start);
        return ans;
    }
    
    private boolean isPossibleToReach(int curIndex) {
        
        if (curIndex < 0 || curIndex >= arr.length)
            return false;
        
        if (this.visited.contains(curIndex) == true)
            return false;
        
        if (this.arr[curIndex] == 0) return true;
        
        this.visited.add(curIndex);
        
        boolean ans = false;
        if (isPossibleToReach(curIndex + this.arr[curIndex]) == true) ans = true;
        else if (isPossibleToReach(curIndex - this.arr[curIndex]) == true) ans = true;
        
        // backtrack
        this.visited.remove(curIndex);
        
        return ans;
    }
}

File: ..solutions/1306. Jump Game III.cpp
class Solution {
    vector<bool> visited;
    public:
    bool canReach(vector<int>& nums, int start) {
        int n = nums.size();
        visited.resize(n+2);
        queue<int> q;
        
        q.push(start);
        
        while(!q.empty()) {
            start = q.front();
            q.pop();
            
            // go start - nums[start]
            {if(start - nums[start] >= 0 && !visited[start]) {
                if(start - nums[start] == 0 || nums[start] == 0) {
                    return true;
                }else if(start - nums[start] > 0) {
                    visited[start] = true;
                    q.push(start-nums[start]);
                }
            }else {continue;}}
            // go start + nums[start]
            {
                if(start + nums[start] < (int) nums.size()  && !visited[start]) {
                if(start + nums[start] == 0 || nums[start] == 0 ) {
                    return true;
                }else if (start - nums[start] > 0) {
                    visited[start] = true;
                    q.push(start+nums[start]);
                }
            }else {continue;}
            }
        }
        return false;
    
        // bool ans = dfs(nums, start);
        // return ans;
    }
​
    bool dfs(vector<int> &nums, int start) {
        if(start < 0 || start >= (int) nums.size() || visited[start]) {
            return false;
        }
        visited[start] = true;
        if(nums[start] == 0){
            return true;
        }
        bool ans1 = dfs(nums, start + nums[start]);
        bool ans2 = dfs(nums, start - nums[start]);
        return ans1 || ans2;
    }
};
